### ISSUE TITLE
Elements (Chunks, Snippets, etc.) Check if there is a more recent update

### ISSUE THREAD

**Issue by pepebe** [modxcms/revolution#13626]
### Summary
It happens from time to time, especially if you are working on multiple screens. For some reason you have two open tabs displaying the same chunk. They might be on the same window or in different windows, it doesn't really matter. 

You work in the first tab for two hours, save your work and close it. Then you treat yourself to a cup of strong coffee and after returning to your desk, you somehow save the old chunk over your hard work. Sometimes you might notice it at once and you do something about it, sometimes you don't notice and do nothing.

###  Current workaround 
Install versionx or some other kind ov versioning solution.

### Expected behavior
An element should know its last update timestamp BEFORE saving the current version, compare that with the current status in the database and if there is a more recent version already present ask for confirmation.

**Comment by mindeffects** [comment]
Yes, please! But this mechanism is already working for resources so there shouldn't be a need your another solution. All the stuff we need is already on board of MODX. (In case you are using static files, well, this is your "local" problem. ;-) )

**Comment by OptimusCrime** [comment]
The system with locks for resources is very complex. I think another (easier) solution could be used for elements. When the data is fetched from the processor, we compile a checksum (md5 or something) which is used to verify the element state. Upon saving the element, we send the old checksum, re-calculate the current element state (stored in the database) and see if they match. If they differ, then the element state has changed since we fetched the data from the processor. This can prompt the user with a warning, giving the option to cancel saving, or overwrite the changes. I think this should be pretty doable. The problem is how to calculate the cheksum. Just the content/code/markup should be easy enough, but should we include the description too? Property sets?

**Comment by mindeffects** [comment]
Why not a simple check-in-out? If the element is fetched for editing within the manager it is checked in (usename&|timestamp is stored). If the element gets saved and(!) closed it will get checked out. If a second attempt opens the element meanwhile, it'll get a warning and save is disabled.
Maybe to simple?

**Comment by OptimusCrime** [comment]
This will require new fields in all the element tables. This is something that has been discussed for some time, so this might be a good idea. However, this approach will give false positives if you save the element multiple times (without doing any actual changes). Also, checking for user might be error prone, if executed via a processor. So it would have to look at the timestamp only.

**Comment by rthrash** [comment]
Resource locking in Revo is mediocre at best. They expire after an arbitrary time, which I think the default is 5 minutes. Further, I am not certain that closing expires the lock either. Locks should be a lot simpler and much less magic—they're not reliable IMO.

**Comment by pepebe** [comment]
Oh, I just took for granted that every element has properties like createdby, editedby, editedon, etc. This feels very wrong. Did Evo have them?

I think that all elements should have those fields. It would be a first step to improve a lot of things. It would also help with forensics, etc. Who moved my cheese....

How difficult would it be to add this to the current processors? I guess most of the necessary functions are already used in other places.

**Comment by Pathologic** [comment]
@pepebe Evo elements have createdon, editedon and disabled. They were implemented for consistency not long ago.

**Comment by pepebe** [comment]
Great feature!

Who changed that chunk? When was it last edited, who created it? From my point of view these are valid questions that deserve to be answered.

I can remember a VERY emotional debate whether "createdon" and "createdby" should be editable or not (opt-in). For some reason this doesn't bother anyone for all the other elements.

**Comment by sdrenth** [comment]
I'm going to try and fix this today #modxbughunt
